File: legacy/Shared/ProgressInfo.swift
import Foundation

/// Progress information reported by the worker process
public struct ProgressInfo: Codable, Sendable {
    /// Current frame being processed
    public let frame: Int

    /// Total frames in the video
    public let totalFrames: Int

    /// Current processing speed in frames per second
    public let fps: Double

    /// Estimated time remaining in seconds
    public let eta: TimeInterval

    /// Progress as a fraction (0.0 to 1.0)
    public var progress: Double {
        guard totalFrames > 0 else { return 0 }
        return Double(frame) / Double(totalFrames)
    }

    /// Progress as a percentage (0 to 100)
    public var percentComplete: Int {
        Int(progress * 100)
    }

    /// Formatted ETA string (e.g., "1h 23m 45s")
    public var etaFormatted: String {
        guard eta > 0, eta.isFinite else { return "--" }

        let hours = Int(eta) / 3600
        let minutes = (Int(eta) % 3600) / 60
        let seconds = Int(eta) % 60

        if hours > 0 {
            return String(format: "%dh %02dm %02ds", hours, minutes, seconds)
        } else if minutes > 0 {
            return String(format: "%dm %02ds", minutes, seconds)
        } else {
            return String(format: "%ds", seconds)
        }
    }

    /// Formatted FPS string
    public var fpsFormatted: String {
        guard fps > 0, fps.isFinite else { return "-- fps" }
        return String(format: "%.1f fps", fps)
    }

    public init(frame: Int, totalFrames: Int, fps: Double, eta: TimeInterval) {
        self.frame = frame
        self.totalFrames = totalFrames
        self.fps = fps
        self.eta = eta
    }
}

/// Messages sent from worker to main app via stdout
public enum WorkerMessage: Codable, Sendable {
    case progress(ProgressInfo)
    case log(LogMessage)
    case error(String)
    case complete(success: Bool, outputPath: String?)

    private enum CodingKeys: String, CodingKey {
        case type
        case frame, totalFrames, fps, eta
        case level, message
        case success, outputPath
    }

    private enum MessageType: String, Codable {
        case progress
        case log
        case error
        case complete
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(MessageType.self, forKey: .type)

        switch type {
        case .progress:
            let frame = try container.decode(Int.self, forKey: .frame)
            let totalFrames = try container.decode(Int.self, forKey: .totalFrames)
            let fps = try container.decode(Double.self, forKey: .fps)
            let eta = try container.decode(TimeInterval.self, forKey: .eta)
            self = .progress(ProgressInfo(frame: frame, totalFrames: totalFrames, fps: fps, eta: eta))

        case .log:
            let level = try container.decodeIfPresent(String.self, forKey: .level) ?? "info"
            let message = try container.decode(String.self, forKey: .message)
            self = .log(LogMessage(level: LogLevel(rawValue: level) ?? .info, message: message))

        case .error:
            let message = try container.decode(String.self, forKey: .message)
            self = .error(message)

        case .complete:
            let success = try container.decode(Bool.self, forKey: .success)
            let outputPath = try container.decodeIfPresent(String.self, forKey: .outputPath)
            self = .complete(success: success, outputPath: outputPath)
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)

        switch self {
        case .progress(let info):
            try container.encode(MessageType.progress, forKey: .type)
            try container.encode(info.frame, forKey: .frame)
            try container.encode(info.totalFrames, forKey: .totalFrames)
            try container.encode(info.fps, forKey: .fps)
            try container.encode(info.eta, forKey: .eta)

        case .log(let log):
            try container.encode(MessageType.log, forKey: .type)
            try container.encode(log.level.rawValue, forKey: .level)
            try container.encode(log.message, forKey: .message)

        case .error(let message):
            try container.encode(MessageType.error, forKey: .type)
            try container.encode(message, forKey: .message)

        case .complete(let success, let outputPath):
            try container.encode(MessageType.complete, forKey: .type)
            try container.encode(success, forKey: .success)
            try container.encodeIfPresent(outputPath, forKey: .outputPath)
        }
    }
}

/// Log message from worker
public struct LogMessage: Codable, Sendable {
    public let level: LogLevel
    public let message: String
    public let timestamp: Date

    public init(level: LogLevel, message: String, timestamp: Date = Date()) {
        self.level = level
        self.message = message
        self.timestamp = timestamp
    }
}

/// Log levels
public enum LogLevel: String, Codable, Sendable {
    case debug
    case info
    case warning
    case error
}

/// Processing state machine
public enum ProcessingState: Equatable, Sendable {
    case idle
    case preparingJob
    case processing(progress: Double)
    case cancelling
    case completed(success: Bool)
    case failed(error: String)

    public var isActive: Bool {
        switch self {
        case .preparingJob, .processing, .cancelling:
            return true
        default:
            return false
        }
    }

    public var canCancel: Bool {
        switch self {
        case .preparingJob, .processing:
            return true
        default:
            return false
        }
    }
}

File: worker/src/models/progress_info.rs
//! Progress information and worker messages.

use serde::{Deserialize, Serialize};
use chrono::{DateTime, Utc};

/// Progress information reported by the worker process.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProgressInfo {
    /// Current frame being processed
    pub frame: i32,

    /// Total frames in the video
    pub total_frames: i32,

    /// Current processing speed in frames per second
    pub fps: f64,

    /// Estimated time remaining in seconds
    pub eta: f64,
}

impl ProgressInfo {
    /// Create a new progress info.
    pub fn new(frame: i32, total_frames: i32, fps: f64, eta: f64) -> Self {
        Self {
            frame,
            total_frames,
            fps,
            eta,
        }
    }

    /// Progress as a fraction (0.0 to 1.0).
    pub fn progress(&self) -> f64 {
        if self.total_frames <= 0 {
            return 0.0;
        }
        (self.frame as f64) / (self.total_frames as f64)
    }

    /// Progress as a percentage (0 to 100).
    pub fn percent_complete(&self) -> i32 {
        (self.progress() * 100.0) as i32
    }

    /// Formatted ETA string (e.g., "1h 23m 45s").
    pub fn eta_formatted(&self) -> String {
        if self.eta <= 0.0 || !self.eta.is_finite() {
            return "--".to_string();
        }

        let total_secs = self.eta as i64;
        let hours = total_secs / 3600;
        let minutes = (total_secs % 3600) / 60;
        let seconds = total_secs % 60;

        if hours > 0 {
            format!("{}h {:02}m {:02}s", hours, minutes, seconds)
        } else if minutes > 0 {
            format!("{}m {:02}s", minutes, seconds)
        } else {
            format!("{}s", seconds)
        }
    }

    /// Formatted FPS string.
    pub fn fps_formatted(&self) -> String {
        if self.fps <= 0.0 || !self.fps.is_finite() {
            return "-- fps".to_string();
        }
        format!("{:.1} fps", self.fps)
    }
}

/// Messages sent from worker to main app via stdout.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum WorkerMessage {
    /// Progress update
    Progress {
        frame: i32,
        #[serde(rename = "totalFrames")]
        total_frames: i32,
        fps: f64,
        eta: f64,
    },

    /// Log message
    Log {
        level: String,
        message: String,
    },

    /// Error message
    Error {
        message: String,
    },

    /// Job completion
    Complete {
        success: bool,
        #[serde(rename = "outputPath", skip_serializing_if = "Option::is_none")]
        output_path: Option<String>,
    },
}

impl WorkerMessage {
    /// Create a progress message.
    pub fn progress(info: &ProgressInfo) -> Self {
        WorkerMessage::Progress {
            frame: info.frame,
            total_frames: info.total_frames,
            fps: info.fps,
            eta: info.eta,
        }
    }

    /// Create a log message.
    pub fn log(level: LogLevel, message: &str) -> Self {
        WorkerMessage::Log {
            level: level.as_str().to_string(),
            message: message.to_string(),
        }
    }

    /// Create an error message.
    pub fn error(message: &str) -> Self {
        WorkerMessage::Error {
            message: message.to_string(),
        }
    }

    /// Create a completion message.
    pub fn complete(success: bool, output_path: Option<&str>) -> Self {
        WorkerMessage::Complete {
            success,
            output_path: output_path.map(String::from),
        }
    }
}

/// Log message from worker.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LogMessage {
    pub level: LogLevel,
    pub message: String,
    pub timestamp: DateTime<Utc>,
}

impl LogMessage {
    pub fn new(level: LogLevel, message: &str) -> Self {
        Self {
            level,
            message: message.to_string(),
            timestamp: Utc::now(),
        }
    }
}

/// Log levels.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum LogLevel {
    Debug,
    Info,
    Warning,
    Error,
}

impl LogLevel {
    pub fn as_str(&self) -> &'static str {
        match self {
            LogLevel::Debug => "debug",
            LogLevel::Info => "info",
            LogLevel::Warning => "warning",
            LogLevel::Error => "error",
        }
    }
}

/// Processing state machine.
#[derive(Debug, Clone, PartialEq)]
pub enum ProcessingState {
    Idle,
    PreparingJob,
    Processing { progress: f64 },
    Cancelling,
    Completed { success: bool },
    Failed { error: String },
}

impl ProcessingState {
    /// Check if processing is active.
    pub fn is_active(&self) -> bool {
        matches!(
            self,
            ProcessingState::PreparingJob
                | ProcessingState::Processing { .. }
                | ProcessingState::Cancelling
        )
    }

    /// Check if cancellation is possible.
    pub fn can_cancel(&self) -> bool {
        matches!(
            self,
            ProcessingState::PreparingJob | ProcessingState::Processing { .. }
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_progress_info() {
        let info = ProgressInfo::new(500, 1000, 25.0, 20.0);
        assert_eq!(info.progress(), 0.5);
        assert_eq!(info.percent_complete(), 50);
        assert_eq!(info.eta_formatted(), "20s");
        assert_eq!(info.fps_formatted(), "25.0 fps");
    }

    #[test]
    fn test_worker_message_serialization() {
        let msg = WorkerMessage::progress(&ProgressInfo::new(100, 1000, 30.0, 30.0));
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains("\"type\":\"progress\""));
        assert!(json.contains("\"frame\":100"));
    }

    #[test]
    fn test_log_message_serialization() {
        let msg = WorkerMessage::log(LogLevel::Info, "Test message");
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains("\"type\":\"log\""));
        assert!(json.contains("\"level\":\"info\""));
    }
}
